NMS-10273: Add support for Cisco Syslog Message formats to the RadixTreeSyslogParser

mattixtech opened the pull request:
* JIRA: http://issues.opennms.org/browse/NMS-10273

Our [continuous integration system](http://bamboo.internal.opennms.com:8085) will test and verify your changes.

j-white commented on features/events/syslog/src/main/java/org/opennms/netmgt/syslogd/ParserStageSequenceBuilder.java:
```diff
@@ -715,6 +822,46 @@ public boolean equals(Object o) {
+					return AcceptResult.COMPLETE_WITHOUT_CONSUMING;
+				}
+			}
+			// TODO: Make this more efficient?
```
Is there a way to make this more efficient? `c` is a single character, so comparing it against a set of known white-space characters (which would be truncated by `trim()`) may be faster.

j-white commented on features/events/syslog/src/main/java/org/opennms/netmgt/syslogd/ParserStageSequenceBuilder.java:
```diff
@@ -152,6 +172,38 @@ private void addStage(ParserStage stage) {
+			case IPADDRESS:
+				return ipV4OrV6AddressMatcher;
+			default:
+				throw new IllegalArgumentException("Unsupported pattern.");
```
I would include the pattern in the exception message.

j-white commented on features/events/syslog/src/main/resources/org/opennms/netmgt/syslogd/grok-patterns.txt:
```diff
@@ -78,3 +81,15 @@
+
+# Note: This pattern '%{STRING:message}' should always come last as it will greedily match anything. If a pattern is
+# placed after this one it will never be matched since this pattern matches anything.
+# TODO: Update this to take any string beginning with '%' i.e. %{BEGINSWITH_%:message}
```
It may not be a bad thing to leave this here without matching `%` first. This way the message won't be dropped, and given the new behavior we've introduced it will be associated with the current time, and the source host.

mattixtech commented:
Addressed your feedback. Also included a fix for a corner case I didn't originally notice (where the hostname is the last pattern in a grok string, that is handled in its own section).

j-white commented on features/events/syslog/src/main/java/org/opennms/netmgt/syslogd/GrokParserStageSequenceBuilder.java:
```diff
@@ -198,7 +208,17 @@
+		/**
+		 * Sequence number.
+		 */
+		parmSequenceNum,
```
Is there an easy way to make the parameter naming generic so that we don't have to hard-code the parameter names?

mattixtech replied:
Could use the text form the grok pattern as the key for the map. That's easy for the map.put(). For the get() you would need to know the literal string for the parameter you want to get since you could no longer reference it here. Probably safe to assume those pattern names will never change and if so then we can drop the grok pattern -> key mapping and get rid of this enum. There's also no getters at the moment to worry about. So yes, I'll change this.

j-white commented:
Thanks for the update 👍
